### PR TITLE
[WIP] OS Infra integration

### PR DESCRIPTION
**Pull request by mfalesni**
cc @petrblaho , @mcornea

{{pytest: cfme/tests/cloud/test_providers.py cfme/tests/infrastructure/test_providers.py -v -k provider_crud --use-provider tripleo --long-running }}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 75bdf0c7ad4ddec8fdf85f93f0cb13c8d162711a:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./utils/mgmt_system/openstack_infra.py:2:1: F401 'heatclient' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 34ea8ac1ca775cdf046c4590efa8b360a3d41dce:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./utils/mgmt_system/openstack_infra.py:9:1: F401 'current_version' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 5b6b4a5537563c3895dcd7da8a36cdb5fa01237f:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 4b97367cab9a1e855b7b2ac139bbce1f785b0858:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 7a49bc5a50ec8d46a8de0fd3298aa6413513394f:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by mfalesni** [comment]
I believe this could be merged for starter, but there still is a networking problem as when the appliance is located in the same x.x.x. subnet, it does not see the tripleo range, probably the proxy arp does not work in this case? @seandst @psav ?

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 09aaf3c33c1be51d6661a0c768d10e0e53100989:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 89b3b1c7bfe6f98920464a83f700642f529462cd:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 293fefcdfbee8222e3c4def79ec1e2500249a996:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 3ff6e03f04f2838431adaecce45cbe7d142eb99b:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit b8411084b246ab8c820204b2d1a7f8dbac0248a3:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 854ffbfcc0d37c190e58cd82d8f945274b9e5f86:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./fixtures/pytest_store.py:29:1: F401 'on_signal' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 7230459cacdc3487170cdbdf1400e07e3de3f2ac:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./fixtures/pytest_store.py:29:1: F401 'on_signal' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by mfalesni** [comment]
I found out the logic behind UI changes is not momentary, dependent on one event. Rather than scavenge the data from database, I am working on using non-text informations for some UI things. First one is ability to specify a link (starting with /) to look for in the menu, that circumvents the need of picking the correct link title in the scond level. I also created a function that does a lookup for the infoblock item name by the icon in the value, taht circumvents picking in the provider summary. The Control stuff and so seem to just change everywhere not mattering whether there is a OS infra provider or isn't.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 0ad38ae11015f12c42c92341d5536481e7849051:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/cloud/provider.py:31:1: F401 'provider_factory' imported but unused
./fixtures/pytest_store.py:29:1: F401 'on_signal' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 11337d842ebcb6ee9d2020b38eda2f122f0dbcb4:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 0aee3396a51ca60b48b038011f2a90751f0a1dd3:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit b5bce51b6f5f13f691ed42ae72233ed6c11fbd79:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit f96774b1927cfd201f9abd57d07f06b42c7477ed:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/infrastructure/provider.py:11:1: F401 'copy' imported but unused
./cfme/infrastructure/provider.py:25:1: F401 'TabStripForm' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 750f70fb2c0ab52684af272ae492ed7055752127:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/infrastructure/provider.py:11:1: F401 'copy' imported but unused
./cfme/infrastructure/provider.py:25:1: F401 'TabStripForm' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 4698e1fdf6712656477aa071569590fce67e367c:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/infrastructure/provider.py:11:1: F401 'copy' imported but unused
./cfme/infrastructure/provider.py:26:1: F401 'TabStripForm' imported but unused
./cfme/infrastructure/provider.py:31:1: F401 'deferred_verpick' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit eb067eee23c886661b797282f591d9971290ec52:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/infrastructure/provider.py:11:1: F401 'copy' imported but unused
./cfme/infrastructure/provider.py:26:1: F401 'TabStripForm' imported but unused
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8d10320deb79009e4024f1026231cb7c9f84825e:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 69f2ceb9ccb601409d67647d071c33b00227ab98:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ea628a540c4d62f05bf3020b4162813138686ca0:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a2a98bdf43c28c371ee5ea2803a048f8dc56bf42:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2c8fd9b60a32e5a49d4db164ee6016aea64a9b9c:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

File: cfme/infrastructure/provider.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ def num_host(self, db=True):
                             .filter(ext_management_systems.name == self.name))
             return len(hostlist)
         else:
-            return int(self.get_detail("Relationships", "Hosts"))
+            return int(self.get_detail("Relationships", "host.png", use_icon=True))
```
Comment (psav) [inline cfme/infrastructure/provider.py]: Does ```Hosts``` no longer work now?

File: utils/providers.py [hunk 2]
```diff
@@ -286,6 +287,15 @@ def setup_provider(provider_key, validate=True, check_existing=True):
...
     provider = get_crud(provider_key)
+    if hasattr(provider, "infra_provider"):
+        if provider.infra_provider not in {None, False}:
+            from cfme.infrastructure.provider import OpenstackInfraProvider
+            if isinstance(provider.infra_provider, OpenstackInfraProvider):
+                infra_provider_name = provider.infra_provider.name
+            else:
+                infra_provider_name = str(provider.infra_provider)
+            infra_provider_key = get_provider_key(infra_provider_name)
+            setup_provider(infra_provider_key, validate=validate, check_existing=check_existing)
```
Comment (psav) [inline utils/providers.py]: Just so I understand.......what are we setting up here?

Comment (mfalesni) [inline utils/providers.py]: I don't remember why exactly was I putting it here. To be consistent with the setup_provider parameters? Not sure.

File: cfme/infrastructure/provider.py (hunk 1 above)
Comment (mfalesni) [inline cfme/infrastructure/provider.py]: It changes to `Hosts / Nodes` when you have at least one OS Infra provider. And I was not able to reliably determine the moment when to switch so going with the path of least resistance.

Comment (psav) [inline cfme/infrastructure/provider.py]: Makes sense, just frustrating we have to go to these lengths :)

File: utils/providers.py (hunk 2 above)
Comment (psav) [inline utils/providers.py]: I'm not sure I understand it that's all, we may _not_ get a crud object? or we may get a non Infra crud object, but with an infra_provider attr??

Comment (mfalesni) [inline utils/providers.py]: This sets up the underlying infra provider if that is set in the object, so the infrastructure object can be selected during the cloud provider setup in the form. It then returns to provider that you wanted.

Comment (psav) [inline utils/providers.py]: Figured it was something like that

**Comment by dajoRH** [comment]
Lint report for commit b21136b0cdcedb729705668971a4f537f7533f4d:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by mfalesni** [comment]
WIP'ed just to not let the PR tester pick it up after squash.

**Comment by dajoRH** [comment]
Lint report for commit 0ce4bd8ae1c97d7765c373de387e057a4e597346:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by mfalesni** [comment]
@psav merge?

**Pull request merged**